src/cudamatrix/cudnn-convolution-test.cc

```diff
@@ -0,0 +1,120 @@
+// cudamatrix/cudnn-convolution-test.cc
+
+// Copyright 2016  Johns Hopkins University (author: Daniel Povey)
+//           2016  Yiming Wang
+
+// See ../../COPYING for clarification regarding multiple authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY IMPLIED
+// WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABLITY OR NON-INFRINGEMENT.
+// See the Apache 2 License for the specific language governing permissions and
+// limitations under the License.
+#if HAVE_CUDA == 1 && HAVE_CUDNN == 1
+#include <iostream>
+
+#include "base/kaldi-common.h"
+#include "util/common-utils.h"
+#include "cudamatrix/cudnn-convolution.h"
+
+using namespace kaldi;
+
+
+namespace kaldi {
+
+template<typename Real>
+static void UnitTestCudnnConvolutionForward() {
+  CuDnnConvolution<Real> cudnnConv;
+  size_t nbDimsX = 5;
+  MatrixIndexT dimX[] = {100, 20, 300, 200, 100};
+  MatrixIndexT sizeX = 1;
+  for (int32 i = 0; i < nbDimsX; i++) {
+    sizeX *= dimX[i];
+  } 
+  MatrixIndexT strideX[] = {20 * 300 * 200 * 100, 300 * 200 * 100, 200 * 100, 100, 1};
+  cudnnTensorDescriptor_t xDesc;
+  cudnnConv.InitializeTensorDescriptor(nbDimsX, dimX, strideX, xDesc);
+  CuMatrix<Real> x(1, sizeX);
```

Comment:
I see what you're trying to do here. You're avoiding creating a matrix to represent the tensor since a matrix only has two dimensions anyway, and it uses cudaMallocPitched, which may or may not make the stride equal to the number of columns. Dan recently made a change so that you can ensure that stride equals the number of columns, via this constructor in cu-matrix.h:

```
  /// Constructor with memory initialisation                                                             
  CuMatrix(MatrixIndexT rows, MatrixIndexT cols,
           MatrixResizeType resize_type = kSetZero,
           MatrixStrideType stride_type = kDefaultStride) {
    Resize(rows, cols, resize_type, stride_type);
  }
```

where you set stride_type = kStrideEqualNumCols.

Since the nnet3 component interface uses matrices where the row dimensions represents the batch, you will want to depend on this constructor having been used to create the input and output batch matrices. We already do in RepeatedAffineComponent.